File: src/common/web/features/Features.ts
import logging from "../core/logging/Logging";
import { type Constructable } from "../utils/Types";
import { DataManagementPlanFeature } from "./DataManagementPlanFeature";
import { MetadataFeature } from "./MetadataFeature";
import { ProjectFeature } from "./ProjectFeature";
import { findPanelLoader, type ProjectFeaturePanelLoaders } from "./ProjectFeaturePanelLoader";
import { ProjectFeaturesCatalog } from "./ProjectFeaturesCatalog";
import { SummaryFeature } from "./SummaryFeature";

/**
 * Registers all available project features.
 *
 * When adding a new project feature, always register it here.
 *
 * @param panelLoaders - Optional panel loaders to provide panels for the various features.
 */
export function registerProjectFeatures(panelLoaders?: ProjectFeaturePanelLoaders): void {
    interface ConstructableFeature extends Constructable<ProjectFeature> {
        FeatureID: string;
    }

    function registerFeature(feature: ConstructableFeature): void {
        ProjectFeaturesCatalog.registerItem(feature.FeatureID, new feature(findPanelLoader(feature.FeatureID, panelLoaders)));
    }

    // New project features go here; the order in which they are added also define the order of their panels
    registerFeature(MetadataFeature);
    registerFeature(DataManagementPlanFeature);
    registerFeature(SummaryFeature);

    // Print all available features for debugging purposes
    const names = Object.keys(ProjectFeaturesCatalog.items).map((item) => `${ProjectFeaturesCatalog.items[item].displayName} (${item})`);
    logging.debug(`Registered project features: ${names.join("; ")}`);
}

File: src/common/py/features/features.py
import typing

from .data_management_plan_feature import DataManagementPlanFeature
from .metadata_feature import MetadataFeature
from .project_features_catalog import ProjectFeaturesCatalog
from .summary_feature import SummaryFeature
from ..core.logging import debug


def register_project_features() -> None:
    """
    Registers all available project features.

    When adding a new project feature, always register it here.
    """

    def register_feature(feat) -> None:
        ProjectFeaturesCatalog.register_item(feat.FEATURE_ID, feat())

    #  New project features go here
    register_feature(MetadataFeature)
    register_feature(DataManagementPlanFeature)
    register_feature(SummaryFeature)

    # Print all available features for debugging purposes
    names: typing.List[str] = []
    for feature_id, feature in ProjectFeaturesCatalog.items():
        names.append(f"{feature.display_name} ({feature_id})")
    debug(f"Registered project features: {'; '.join(names)}")
